jellyfish/_jellyfish.py

```diff
@@ -111,6 +113,133 @@ def _jaro_winkler(ying, yang, long_tolerance, winklerize):
+        Y = (keyboard_cartesian_spaces[a]['y']-keyboard_cartesian_spaces[b]['y'])**2
+        Z = (keyboard_cartesian_spaces[a]['z']-keyboard_cartesian_spaces[b]['z'])**2
+        cost = sqrt(X+Y+Z)
+    
+    # normalize over threshold of tolerable error, this makes substitution in this region always 
+    # less costly than insertion or deletion.
+    normalized_cost = cost/sqrt(7.68)
+    return cost
+
+def _encapsulate_ops(operations):
+
+    op_names = ['insertion','deletion','substitution','match']
+    ops = dd(int)
+    for op in operations:
+        ops[op]+=1
+    for op_name in op_names:
+        if op_name in ops.keys():
+            pass
+        else:
+            ops[op_name]=0
+    return ops
+
+def _backtrace(target, source, rows):
+
+    i, j = len(target), len(source)
+    edits = []
+
+    while(not (i == 0  and j == 0)):
+        
+        prev_cost = rows[i][j]
+     
+        neighbors = []
+     
+        if(i!=0 and j!=0):
+          neighbors.append(rows[i-1][j-1])
+        if(i!=0):
+          neighbors.append(rows[i-1][j])
+        if(j!=0):
+          neighbors.append(rows[i][j-1])
+     
+        min_cost = min(neighbors)
+ 
+        if(min_cost == prev_cost):
+          i, j = i-1, j-1
+          edits.append('match')
+        elif(i!=0 and j!=0 and min_cost == rows[i-1][j-1]):
+          i, j = i-1, j-1
+          edits.append('substitution')
+        elif(i!=0 and min_cost == rows[i-1][j]):
+          i, j = i-1, j
+          edits.append('deletion')
+        elif(j!=0 and min_cost == rows[i][j-1]):
+          i, j = i, j-1
+          edits.append('insertion')
+    
+    edits.reverse() 
+    return edits
+
+# v basic Minimum Edit Distance function with optional added tracking of operations
+def qwerty_levenshtein(target, source):
+
+    len_target, len_source = len(target),len(source)
+    
+    distance = [[0 for i in range(len_source+1)] for j in range(len_target+1)]
+    
+    for i in range (1,len_target+1):
+        distance[i][0] = distance[i-1][0] + 1
+    for j in range(1, len_source+1):
+        distance[0][j] = distance[0][j-1] + 1
+    for i in range (1,len_target+1):
+        for j in range(1, len_source+1):
+           
+            try:
+                sub_cost = _qwerty_sub_cost(source[j-1],target[i-1])
+            except KeyError, e:
+                cause=e.args[0]
+                if cause in source:
+                    new_source = re.sub(cause,'',source)
+                    return qwerty_levenshtein(target, new_source)
+
+                if cause in target:
+                    new_target=re.sub(cause,'',target)
+                    return qwerty_levenshtein(new_target, source)
+            
+            insert = distance[i-1][j]+1
+            delete = distance[i][j-1]+1
+            substitute = distance[i-1][j-1]+sub_cost
+            
+            distance[i][j] = min(insert,delete,substitute)
+
+            if i and j and target[i]==source[j-1] and target[i-1]==source[j]:
```

Comment:
IndexError here, should be

`if i and j and target[i-1] == source[j-2] and target[i-2] == source[i-1]`